opencl/btx_clinterval_callbacks.cpp

```diff
@@ -0,0 +1,354 @@
+                                         const char *event_class_name,
+                                         const char *hostname, int64_t vpid,
+                                         uint64_t vtid, cl_int cl_rval) {
+  send_host_message(btx_handle, usr_data, ts, event_class_name, hostname, vpid,
+                    vtid, clRvalIsError(cl_rval));
+}
+
+static void traffic_entry_callback(void *btx_handle, void *usr_data, int64_t ts,
+                                   const char *event_class_name,
+                                   const char *hostname, int64_t vpid,
+                                   uint64_t vtid, size_t size) {
+  // save traffic size and entry ts for use in exit callback
+  auto state = static_cast<data_t *>(usr_data);
+  hpt_t key{hostname, vpid, vtid};
+  state->entry_state.set_ts(key, ts);
+  state->entry_state.set_data(key, size);
+  // Note: flag to deal with traffic exit calls being hard to match on
+  state->last_enqueue_entry_is_traffic[key] = true;
+}
+
+// Note: handles traffic exits, but can't match easily so uses flag to
+// distinguish traffic vs other clEnqueue exits
+static void enqueue_exit_callback(void *btx_handle, void *usr_data, int64_t ts,
+                                  const char *event_class_name,
+                                  const char *hostname, int64_t vpid,
+                                  uint64_t vtid, cl_int errcode_ret_val) {
+  auto state = static_cast<data_t *>(usr_data);
+  hpt_t key{hostname, vpid, vtid};
+  if (!state->last_enqueue_entry_is_traffic[key]) {
+    return;
+  }
+  // clear flag
+  state->last_enqueue_entry_is_traffic[key] = false;
+
+  // Note: don't emit a traffic event if there was an error
+  if (clRvalIsError(errcode_ret_val)) {
+    return;
+  }
+
+  std::string event_class_name_stripped =
+      strip_event_class_name_exit(event_class_name);
+  const int64_t entry_ts = state->entry_state.get_ts(key);
+  auto size = state->entry_state.get_data<size_t>(key);
+  btx_push_message_lttng_traffic(btx_handle, hostname, vpid, vtid, entry_ts,
+                                 BACKEND_OPENCL,
+                                 event_class_name_stripped.c_str(), size);
+}
+
+static void get_device_ids_exit_callback(void *btx_handle, void *usr_data, int64_t ts,
+                                         const char *event_class_name,
+                                         const char *hostname, int64_t vpid,
+                                         uint64_t vtid, cl_int errcode_ret_val,
+                                         cl_uint num_devices_val,
+                                         cl_device_id *devices_vals) {
+  auto state = static_cast<data_t *>(usr_data);
+  if ((devices_vals != nullptr) && (errcode_ret_val == CL_SUCCESS)) {
+    for (unsigned int i=0; i < num_devices_val; i++) {
+      const thapi_device_id d = reinterpret_cast<thapi_device_id>(devices_vals[i]);
+      state->device_to_root_device[hp_device_t(hostname, vpid, d)] = d;
```

Comment:
can remove `hp_device_t` cast an rely of auto deduction

Reply:
it works! Get confused about which cases it actually works for...